File: app/src/main/java/com/marknkamau/justjava/ui/signup/SignUpActivityPresenter.java
package com.marknkamau.justjava.ui.signup;

import android.content.SharedPreferences;
import android.os.Handler;

import com.marknkamau.justjava.models.UserDefaults;
import com.marknkamau.justjava.utils.Constants;
import com.marknkamau.justjava.utils.FirebaseAuthUtils;
import com.marknkamau.justjava.utils.FirebaseDBUtil;

class SignUpActivityPresenter {
    private SignUpActivityView activityView;
    private SharedPreferences sharedPreferences;

    SignUpActivityPresenter(SignUpActivityView activityView, SharedPreferences sharedPreferences) {
        this.activityView = activityView;
        this.sharedPreferences = sharedPreferences;
    }

    void createUser(final String email, final String password, final String name, final String phone, final String address) {
        activityView.disableUserInteraction();

        FirebaseAuthUtils.INSTANCE.createUser(email, password, new FirebaseAuthUtils.AuthActionListener() {
            @Override
            public void actionSuccessful(String response) {
                signInUser(email, password, name, phone, address);
            }

            @Override
            public void actionFailed(String response) {
                activityView.enableUserInteraction();
                activityView.displayMessage(response);
            }
        });
    }

    private void signInUser(String email, String password, final String name, final String phone, final String address) {
        FirebaseAuthUtils.INSTANCE.signIn(email, password, new FirebaseAuthUtils.AuthActionListener() {
            @Override
            public void actionSuccessful(String response) {
                setUserDisplayName(name, phone, address);
            }

            @Override
            public void actionFailed(String response) {
                activityView.enableUserInteraction();
                activityView.displayMessage(response);
            }
        });
    }

    private void setUserDisplayName(final String name, final String phone, final String address) {
        FirebaseAuthUtils.INSTANCE.setUserDisplayName(name, new FirebaseAuthUtils.AuthActionListener() {
            @Override
            public void actionSuccessful(String response) {
                FirebaseDBUtil.INSTANCE.setUserDefaults(new UserDefaults(name, phone, address), new FirebaseDBUtil.UploadListener() {
                    @Override
                    public void taskSuccessful() {
                        activityView.enableUserInteraction();
                        saveToSharedPreferences(name, phone, address);
                        activityView.displayMessage("Sign up successfully");
                        new Handler().postDelayed(new Runnable() {
                            @Override
                            public void run() {
                                activityView.finishActivity();
                            }
                        }, 500);

                    }

                    @Override
                    public void taskFailed(String response) {
                        activityView.enableUserInteraction();
                        activityView.displayMessage(response);
                    }
                });

            }

            @Override
            public void actionFailed(String response) {
                activityView.enableUserInteraction();
                activityView.displayMessage(response);
            }
        });
    }

    private void saveToSharedPreferences(String name, String phone, String address) {
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putString(Constants.INSTANCE.getDEF_NAME(), name);
        editor.putString(Constants.INSTANCE.getDEF_PHONE(), phone);
        editor.putString(Constants.INSTANCE.getDEF_ADDRESS(), address);

        editor.apply();
    }
}

File: app/src/main/java/com/marknkamau/justjava/ui/signup/SignUpActivityPresenter.kt
package com.marknkamau.justjava.ui.signup

import android.os.Handler
import com.marknkamau.justjava.data.PreferencesRepository

import com.marknkamau.justjava.models.UserDefaults
import com.marknkamau.justjava.utils.FirebaseAuthUtils
import com.marknkamau.justjava.utils.FirebaseDBUtil

internal class SignUpActivityPresenter(private val activityView: SignUpActivityView, private val preferences: PreferencesRepository) {

    fun createUser(email: String, password: String, name: String, phone: String, address: String) {
        activityView.disableUserInteraction()

        FirebaseAuthUtils.createUser(email, password, object : FirebaseAuthUtils.AuthActionListener {
            override fun actionSuccessful(response: String) {
                signInUser(email, password, name, phone, address)
            }

            override fun actionFailed(response: String?) {
                activityView.enableUserInteraction()
                activityView.displayMessage(response)
            }
        })
    }

    private fun signInUser(email: String, password: String, name: String, phone: String, address: String) {
        FirebaseAuthUtils.signIn(email, password, object : FirebaseAuthUtils.AuthActionListener {
            override fun actionSuccessful(response: String) {
                setUserDisplayName(name, phone, address)
            }

            override fun actionFailed(response: String?) {
                activityView.enableUserInteraction()
                activityView.displayMessage(response)
            }
        })
    }

    private fun setUserDisplayName(name: String, phone: String, address: String) {
        FirebaseAuthUtils.setUserDisplayName(name, object : FirebaseAuthUtils.AuthActionListener {
            override fun actionSuccessful(response: String) {
                FirebaseDBUtil.setUserDefaults(UserDefaults(name, phone, address), object : FirebaseDBUtil.UploadListener {
                    override fun taskSuccessful() {
                        activityView.enableUserInteraction()
                        preferences.saveDefaults(UserDefaults(name, phone, address))
                        activityView.displayMessage("Sign up successfully")
                        Handler().postDelayed({ activityView.finishActivity() }, 500)

                    }

                    override fun taskFailed(reason: String?) {
                        activityView.enableUserInteraction()
                        activityView.displayMessage(reason)
                    }
                })

            }

            override fun actionFailed(response: String?) {
                activityView.enableUserInteraction()
                activityView.displayMessage(response)
            }
        })
    }
}
